disnake/ext/commands/params.py

```diff
@@ -279,135 +281,159 @@ def decorator(func: CallableT) -> CallableT:
         return decorator
 
 
-class RangeMeta(type):
-    """Custom Generic implementation for Range"""
+@dataclass(frozen=True)
+class _BaseRange(ABC):
+    """Internal base type for supporting ``Range[...]`` and ``String[...]``."""
 
-    @overload
-    def __getitem__(
-        self, args: Tuple[Union[int, EllipsisType], Union[int, EllipsisType]]
-    ) -> Type[int]:
-        ...
+    _allowed_types: ClassVar[Tuple[Type[Any], ...]]
 
-    @overload
-    def __getitem__(
-        self, args: Tuple[Union[float, EllipsisType], Union[float, EllipsisType]]
-    ) -> Type[float]:
-        ...
+    underlying_type: Type[Any]
+    min_value: Optional[Union[int, float]]
+    max_value: Optional[Union[int, float]]
 
-    def __getitem__(self, args: Tuple[Any, ...]) -> Any:
-        a, b = [None if isinstance(x, type(Ellipsis)) else x for x in args]
-        return Range.create(min_value=a, max_value=b)
+    def __class_getitem__(cls, params: Tuple[Any, ...]) -> Self:
+        # deconstruct type arguments
+        if not isinstance(params, tuple):
+            params = (params,)
 
+        name = cls.__name__
 
-class Range(type, metaclass=RangeMeta):
-    """Type depicting a limited range of allowed values.
+        if len(params) == 2:
+            # backwards compatibility for `Range[1, 2]`
+            disnake.utils.warn_deprecated(
+                f"Using `{name}` without an explicit type argument is deprecated, "
+                "as this form does not work well with modern type-checkers. "
+                f"Use `{name}[<type>, {params[0]!r}, {params[1]!r}]` instead.",
+                stacklevel=2,
```

Comment:
this is very wrong when the annotation is commands.String

Reply:
This isn't due to `commands.String`, it's caused by stringified annotations (i.e. `from __future__ import annotations`). `<string>:1` is the eval frame, unfortunately I don't think this is feasible to fix.
The correct source is ~12 frames deep, not 2, going through all the parsing machinery in `params.py` and through `inspect.signature` before being eval'd.

Reply:
Do we have a way to know if the annotation is stringified?

Reply:
probably, but even then, there's just no point in trying to keep the `stacklevel` here in sync with however deep the parameter parsing logic in the library goes. ~~(... ignoring the fact that it differs depending on what called it)~~

Reply:
see #1059 regarding the duplicate warnings

Reply:
> probably, but even then, there's just no point in trying to keep the `stacklevel` here in sync with however deep the parameter parsing logic in the library goes. ~(... ignoring the fact that it differs depending on what called it)~

We should probably add a todo for fixing this for deferred annotations, honestly.

Reply:
sounds good. https://github.com/DisnakeDev/disnake/pull/991/commits/21e78a38aa59e446c44bf587cc088a6e79416abe